### ISSUE TITLE
Complete GPU-CRUD interface

### ISSUE THREAD

**Issue by controllerface** [controllerface/bvge#1]
Access to GPU memory for physics objects requires a complete Create Read Update Delete (CRUD) API as a base layer of functionality upon which many other features can be implemented.

Required methods are purpose built for specific uses rather than being open-ended "blob" objects or something similar. At the moment, there are several create, read, and update methods, but the delete portion is not implemented. 

In order for delete to be implemented, several parallel scan variants will be implemented similar to the existing key bank, candidate, and related prefix-scan based kernels. 

The scan process will operate at the armature level, tallying the counts of all constituent objects (hulls, bones, points, and edges) that will be removed when the armature is deleted. After collecting this information, the backing buffers will be compacted by using the generated scan counts to relocate non-deleted objects to earlier locations in the buffer, filling up the space.

In order to work with the current renderers, the model registry will need to be updated when an associated armature is removed. The armature will need to be removed from the list of tracked models. In addition, because models will relocate to new positions, the model list must also be compacted.

Alternatively, a different design that does not require model registering may be devised.

**Comment by controllerface** [comment]
After doing a bit of prep work on some of the renderers, I noticed an issue with the current design of the circle and crate render batches that will need to be changed in order to support deletes. Luckily, the edge renderer has a design that works well already for a variable number of edges, so my plan is to adopt the same approach for the other renderers, with some changes to account for rendering models instead of edges. 

To make this work, I will need to extend the current armature flags structure from `int` to `int2` with `x` being the current root hull ID, and `y` being the model ID, which will be the same model ID currently used to register a model instance. This will make it possible to write a CL kernel that scans for just a specific model type. The new kernel(s) would work together with the current transforms kernel to extract the necessary world space data to render all the instanced objects.

As part of this, I will also attack the issue of how registering models is currently unable to support deleted models due to the object IDs being mapped. In this new design, I will not need to retain these indices as they will effectively be calculated each frame as needed. With that change, I can make the model register process simply increment a counter, and the delete process would just decrement that same counter, as the only necessary information for rendering is the total count, not the actual object IDs in memory.

**Comment by controllerface** [comment]
I did eventually get to work on this in earnest, but at the moment I am stuck in a state where the deletion of objects is causing issues. I believe my math must be wrong somewhere, but just to give me some space to hash it out, here's the main problem I am seeing:

When I structure the data such that any objects that get deleted are stored _after_ objects that will remain after the delete process, I don't see any issues. The objects seem to be deleted properly and there aren't any visual artifacts (at least I have not seen any in this configuration yet). However, if I put objects to be deleted in the middle of the buffer, after compaction _sometimes_ objects become corrupted and exhibit visual artifacts, such as being "sawn in half" and also some objects indices seem to get messed up, so static objects like the floor crates will move to new locations, sometimes they will start spinning. The geometry seems to still work, as the spinning objects fling colliding objects away, etc. But still, this is not at all the intended outcome.

I have found that when I do tests with very small numbers of objects being deleted, there is a much lower chance of the error occurring, but in almost all configurations, even a few objects _can_ cause the problem. Based on this, my current assumption is that my math and/or methodology for doing the buffer compaction is just wrong, or the operations I am performing in there kernels are not the correct logic to implement what I need. Also, because the problem seems to be intermittent, and there are several places where I am relying on memory barriers, it could be an issue with atomicity/synchronization. 

I also noted in my code that I might want to rescan the objects to generate their boundaries, though initially I put that there only because it may delete an object that is partially on screen. At least on paper, that should be the only shortfall of not rescanning. However, under the current circumstances it may be prudent to try this and see if somehow there is an unexpected side-effect of not doing it.
